File: archive/cdk-java/Pr2266/lib/src/main/java/app/compute/AutoScalingStack.java
package app.compute;

import software.amazon.awscdk.Stack;
import software.amazon.awscdk.StackProps;
import software.amazon.awscdk.services.autoscaling.*;
import software.amazon.awscdk.services.ec2.*;
import software.amazon.awscdk.services.elasticloadbalancingv2.ApplicationTargetGroup;
import software.amazon.awscdk.services.cloudwatch.*;
import software.constructs.Construct;

import java.util.List;

public class AutoScalingStack extends Stack {
    private final AutoScalingGroup autoScalingGroup;
    
    public AutoScalingStack(final Construct scope, final String id, final StackProps props,
                          final Vpc vpc, final SecurityGroup webServerSecurityGroup,
                          final ApplicationTargetGroup targetGroup) {
        super(scope, id, props);
        
        // User data for web server setup
        UserData userData = UserData.forLinux();
        userData.addCommands(
            "yum update -y",
            "yum install -y httpd",
            "systemctl start httpd",
            "systemctl enable httpd",
            "echo '<html><body><h1>Web Server Running</h1><p>Region: " + this.getRegion() + "</p></body></html>' > /var/www/html/index.html",
            "echo 'OK' > /var/www/html/health"
        );
        
        // Launch Template
        LaunchTemplate launchTemplate = LaunchTemplate.Builder.create(this, "WebServerTemplate")
                .machineImage(MachineImage.latestAmazonLinux2())
                .instanceType(InstanceType.of(InstanceClass.T3, InstanceSize.MEDIUM))
                .securityGroup(webServerSecurityGroup)
                .userData(userData)
                .requireImdsv2(true)
                .build();
        
        // Auto Scaling Group
        this.autoScalingGroup = AutoScalingGroup.Builder.create(this, "WebAppASG")
                .vpc(vpc)
                .launchTemplate(launchTemplate)
                .minCapacity(2)
                .maxCapacity(10)
                .desiredCapacity(2)
                .vpcSubnets(SubnetSelection.builder()
                    .subnetType(SubnetType.PRIVATE_WITH_EGRESS)
                    .build())
                .healthCheck(HealthCheck.elb(ElbHealthCheckOptions.builder()
                    .grace(software.amazon.awscdk.Duration.minutes(5))
                    .build()))
                .build();
        
        // Attach to target group
        autoScalingGroup.attachToApplicationTargetGroup(targetGroup);
        
        // Scaling policies
        autoScalingGroup.scaleOnCpuUtilization("CpuScaling",
            CpuUtilizationScalingProps.builder()
                .targetUtilizationPercent(70)
                .cooldown(software.amazon.awscdk.Duration.minutes(3))
                .build());
        
        // Target tracking for ALB requests
        autoScalingGroup.scaleToTrackMetric("RequestCountScaling",
            MetricTargetTrackingProps.builder()
                .metric(targetGroup.metric("RequestCountPerTarget"))
                .targetValue(100)
                .cooldown(software.amazon.awscdk.Duration.minutes(3))
                .build());
    }
    
    public AutoScalingGroup getAutoScalingGroup() {
        return autoScalingGroup;
    }
}

File: archive/cdk-py/Pr3852/lib/autoscaling_stack.py
from aws_cdk import (
    aws_ec2 as ec2,
    aws_autoscaling as autoscaling,
    aws_elasticloadbalancingv2 as elbv2,
    NestedStack,
    Duration,
)
from constructs import Construct


class AutoScalingStack(NestedStack):
    def __init__(
        self,
        scope: Construct,
        construct_id: str,
        *,
        vpc: ec2.Vpc,
        security_group: ec2.SecurityGroup,
        target_group: elbv2.ApplicationTargetGroup,
        **kwargs
    ) -> None:
        super().__init__(scope, construct_id, **kwargs)

        # User data script
        user_data = ec2.UserData.for_linux()
        user_data.add_commands(
            "yum update -y",
            "yum install -y httpd",
            "systemctl start httpd",
            "systemctl enable httpd",
            "echo '<h1>Marketplace Platform</h1>' > /var/www/html/index.html",
            "echo 'OK' > /var/www/html/health",
        )

        # Auto Scaling Group
        self.asg = autoscaling.AutoScalingGroup(
            self,
            "MarketplaceASG",
            vpc=vpc,
            instance_type=ec2.InstanceType.of(
                ec2.InstanceClass.T3, ec2.InstanceSize.LARGE
            ),
            machine_image=ec2.MachineImage.latest_amazon_linux2(),
            security_group=security_group,
            min_capacity=4,
            max_capacity=10,
            vpc_subnets=ec2.SubnetSelection(
                subnet_type=ec2.SubnetType.PRIVATE_WITH_EGRESS
            ),
            user_data=user_data,
        )

        # Attach to target group
        self.asg.attach_to_application_target_group(target_group)

        # Scaling policies
        self.asg.scale_on_cpu_utilization("CPUScaling", target_utilization_percent=70)
